### ISSUE TITLE
Support pyfive as an alternative backend

### ISSUE THREAD

**Issue by shoyer** [h5netcdf/h5netcdf#25]
This will require support for dimension scales (https://github.com/jjhelmus/pyfive/issues/1), but once that is done should be pretty easy, given that pyfive mirrors the h5py API. This will let us read netcdf4 from file-like objects!

CC @jjhelmus

**Comment by kmuehlbauer** [comment]
@shoyer Regarding reading file-like objects, is this still an issue? And second question, should we try to implement this in near future?

**Comment by shoyer** [comment]
h5py can read file-like objects now, so I agree that there is less need for this.

**Comment by shoyer** [comment]
That said, if pyfive ever does support dimension scales, we would still welcome adding it as a backend for h5netcdf.

**Comment by jjhelmus** [comment]
pyfive supports dimension scales and has for some time, jjhelmus/pyfive#14.

**Comment by shoyer** [comment]
Very nice! In that case I'll reopen this in case someone would like to give it a shot. We're all about supporting pure Python implementations of file formats over here at h5netcdf :)

**Comment by kmuehlbauer** [comment]
> And second question, should we try to implement this in near future?

@shoyer  I wasn't clear with my question. What I wanted to ask is, should we try to implement `pyfive` as backend?

**Comment by kmuehlbauer** [comment]
> I'll reopen this in case someone would like to give it a shot.

Your message was first shown after I sent mine. I can give it a try, but do not know if I can dedicate the time in the next months. So if someone is eager to dive into it, please go ahead.

**Comment by shoyer** [comment]
Totally up to you! I think it could be an interesting mini-project, and
might even be useful for cases where it's hard to install HDF5.

**Comment by kmuehlbauer** [comment]
This is like a phone call via a moon relay :grinning:

**Comment by bnlawrence** [comment]
I think there's another blocker right now, at least for "interesting deployments"  (i.e. ones which utilise s3fs): https://github.com/jjhelmus/pyfive/issues/60

**Comment by bnlawrence** [comment]
> I think there's another blocker right now, at least for "interesting deployments" (i.e. ones which utilise s3fs): [jjhelmus/pyfive#60](https://github.com/jjhelmus/pyfive/issues/60)

Though I have a fix now in a forked branch ... we are going to fully test that in another application, but if there is still interest in a pyfive backend, and assuming some variant of my pull requests is accepted there,  is there still  interest here in trying to add a pyfive hdf5 read backend?

**Comment by kmuehlbauer** [comment]
@bnlawrence Yes, thanks, for sure. I've been following the recent activity over at pyfive, great work!

A pure python netcdf4/hdf5 backend would be really neat. Looking forward to this.

**Comment by bnlawrence** [comment]
I'm still making slow progress with this, in doing so I've been adding more "H5PY compatability" to pyfive, even if it's only to raise errors in a way that means where h5py doesn't support something (e.g. enumerated datatypes), the user can still read all the other contents of the file.  Which brings me to the question: I have started an implementation of how one might put pyfive in as a backend ([here](https://github.com/bnlawrence/h5netcdf/tree/pyfive)) with  necessary pyfive modifications [here](https://github.com/bnlawrence/pyfive/tree/h5netcdf)). I am working through a new test, and the obvious thing to decide on now is what to do about enumerations.  For _my_ likely use case, I'm happy to work without enumeration support, are you happy if the new backend, when faced with enumerations simply does [sensible things](https://github.com/bnlawrence/pyfive/issues/8)? Or is the lack of enumeration support a problem for having pyfive as a backend at all?

**Comment by kmuehlbauer** [comment]
@bnlawrence Sorry, this fell through the cracks.

I think it would be great to have office as backend here, even if it's not feature complete.

**Comment by bnlawrence** [comment]
@kmuehlbauer I have a question about something h5netcdf is doing, which I suspect reflects my lack of knowledge of the deep internals of netcdf4 ... I have an error in testing which I can probably fix (will fix), but it arises because of something that h5netcdf is doing that I would rather it didn't do. 

The issue is that the apparently innocuous attempt to get the `dtype` of a h5netcdf variable goes to the file and looks it up via quite a complex routine. There is a similar problem with `shape` which goes via an even more complex routine that ends up looking up things via a dimension_list in the parent file (`_lookup_dimensions`).  Is there any reason why you don't just ask the h5py variable itself what it's `dtype` and `shape` are instead of doing it directly? 

(The reason why I'd rather rely on the variable instance itself is that I have already cached these properties in the variable instance and don't want to go back to the file.)

**Comment by bnlawrence** [comment]
At this point I have a working implementation if anyone else wants to play. You'll need the h5netcdf branch of our [pyfive](https://github.com/NCAS-CMS/pyfive/) and the pyfive branch of our [h5netcdf](https://github.com/NCAS-CMS/h5netcdf) ... All the tests I've implemented (which is just a cut-down version of the fileobj test suite) work.

The next steps will be to clean up our version of pyfive (prune all the branches) and then make a formal pull request upstream. When we have that, we can then open a proper pull request into h5netcdf .... but meanwhile we'd be happy to take criticism of what we have done thus far.

**Comment by bnlawrence** [comment]
(There are still a lot of edge cases around performance and caching to address, but it's already looking pretty good I think).

**Comment by kmuehlbauer** [comment]
@bnlawrence This looks promising. 

Unfortunately I'm struggling a bit to get CompoundData to read (currently only Complex Compounds implemented). Another issue I'm experiencing is that some pyfive np.frombuffer data extraction creates immutable arrays. Not sure if this should be fixed in pyfive or downstream.

It would be great to announce pyfive backend via env variable, like this:

```python
# in h5netcdf.File
if backend is None:
    backend = os.environ.get("H5NETCDF_BACKEND", "h5py")
```

This would help where several downstream packages are involved (xradar, xarray) which might not want to adopt it in fast succession.

**Comment by bnlawrence** [comment]
Three different issues right?

1.  ComplexCompound is a user defined data type, right? I'd have to look into that (I have been tracking complex data types for what seems like forever, but had taken my eye off [the ball](https://github.com/cf-convention/discuss/issues/369#issuecomment-2364562964).)

2. The second is probably more of a problem, and more urgent, but if i have understood the [old ticket](https://github.com/jjhelmus/pyfive/issues/53) that's only an issue when it's a memory mapped array? If so, the fix could be easy? 

3. I am not sure what you mean in the last case? Are you suggesting to remove the keyword option, and just do it all via the environment variable?

**Comment by kmuehlbauer** [comment]
To 3. No, keep the keyword option, but make the env variable available as a fall back (backend=None) case.

**Comment by bnlawrence** [comment]
I was too quick and didn't read it properly, the second looks trivial, as you say, a copy fixes it. Is there anywhere else where that's a problem?

**Comment by bnlawrence** [comment]
> To 3. No, keep the keyword option, but make the env variable available as a fall back (backend=None) case.

If backend is None it currently falls back to h5py. I guess you are suggesting that the fallback position itself could be under environment control. That sounds sensible.

**Comment by kmuehlbauer** [comment]
> If backend is None it currently falls back to h5py. I guess you are suggesting that the fallback position itself could be under environment control. That sounds sensible.

Yes, and if the env var isn't set, it should fallback to h5py/h5pyd.

**Comment by bnlawrence** [comment]
Ok, just pushed up fixes for 3, and hopefully 2. I now have an issue for complex number support in pyfive (https://github.com/NCAS-CMS/pyfive/issues/20), but I think that will be lower priority than getting things into main, after which we can do these sort of things in one-off-issues ... I feel like there is already too much stuff munged together in these changes.

**Comment by bnlawrence** [comment]
We have just issued a comprehensive pull request (https://github.com/jjhelmus/pyfive/pull/68) for a revised `pyfive`, which passes all their original unit tests, and many more that arise from testing as a backend here, and via the [CF-DM](https://ncas-cms.github.io/cfdm/) unit test suite.  We'll wait a bit, and then do a pull request here from our h5netcdf fork - which allows the use of that pyfive as a backend as discussed above.

**Comment by kmuehlbauer** [comment]
Resolved by #279

**Comment by davidhassell** [comment]
Congratulations, Kai - many thanks for your considerable work on this :)

**Comment by valeriupredoi** [comment]
A brilliant Xmas present indeed, very very many thanks @kmuehlbauer 🍺🍺 Happy Xmas to you 🎄 🎅

**Comment by shoyer** [comment]
What an achievement! Well done, and congrats to all.
